src/python_testing/test_testing/TestDecorators.py

```diff
@@ -0,0 +1,292 @@
+        pass
+
+    def step_unknown(self):
+        pass
+
+    def show_prompt(self,
+                    msg: str,
+                    placeholder: Optional[str] = None,
+                    default_value: Optional[str] = None) -> None:
+        pass
+
+
+class TestDecorators(MatterBaseTest):
+    def test_checkers(self):
+        has_onoff = has_cluster(Clusters.OnOff)
+        has_onoff_onoff = has_attribute(Clusters.OnOff.Attributes.OnOff)
+        has_onoff_ontime = has_attribute(Clusters.OnOff.Attributes.OnTime)
+        has_timesync = has_cluster(Clusters.TimeSynchronization)
+        has_timesync_utc = has_attribute(Clusters.TimeSynchronization.Attributes.UTCTime)
+
+        wildcard = get_clusters([0, 1])
+
+        def check_endpoints(f, expect_true, expectation: str):
+            asserts.assert_equal(f(wildcard, 0), expect_true, f"Expected {expectation} == {expect_true} on EP0")
+            asserts.assert_equal(f(wildcard, 1), expect_true, f"Expected {expectation} == {expect_true} on EP1")
+            asserts.assert_false(f(wildcard, 2), f"Expected {expectation} == False on EP2")
+
+        check_endpoints(has_onoff, True, "OnOff Cluster")
+        check_endpoints(has_onoff_onoff, True, "OnOff attribute")
+        check_endpoints(has_onoff_ontime, False, "OnTime attribute")
+        check_endpoints(has_timesync, False, "TimeSynchronization Cluster")
+        check_endpoints(has_timesync_utc, False, "UTC attribute")
+
+    @async_test_body
+    async def test_endpoints(self):
+        has_onoff = has_cluster(Clusters.OnOff)
+        has_onoff_onoff = has_attribute(Clusters.OnOff.Attributes.OnOff)
+        has_onoff_ontime = has_attribute(Clusters.OnOff.Attributes.OnTime)
+        has_timesync = has_cluster(Clusters.TimeSynchronization)
+        has_timesync_utc = has_attribute(Clusters.TimeSynchronization.Attributes.UTCTime)
+
+        all_endpoints = await self.default_controller.Read(self.dut_node_id, [()])
+        all_endpoints = list(all_endpoints.attributes.keys())
+
+        msg = "Unexpected endpoint list returned"
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff)
+        asserts.assert_equal(endpoints, all_endpoints, msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff_onoff)
+        asserts.assert_equal(endpoints, all_endpoints, msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_onoff_ontime)
+        asserts.assert_equal(endpoints, [], msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_timesync)
+        asserts.assert_equal(endpoints, [], msg)
+
+        endpoints = await get_accepted_endpoints_for_test(self, has_timesync_utc)
+        asserts.assert_equal(endpoints, [], msg)
+
+    # This test should cause an assertion because it has pics_ method
+    @per_node_test
+    async def test_whole_node_with_pics(self):
+        pass
+
+    def pics_whole_node_with_pics(self):
```

Comment:
What is `pics_whole_node_with_pics` used for? I cannot seems to find where it is used

Reply:
There is an assertion on the decorator that tests with per_endpoint_test or per_node_test can't have a corresponding pics method, because pics are supplied per endpoint, and these tests require information from the entire node. 

Added a comment on both.